src/sdk/pynni/nni/bohb_advisor/bohb_advisor.py

```diff
@@ -0,0 +1,587 @@
+        ----------
+        data: dict()
+            it has three keys: trial_job_id, event, hyper_params
+            trial_job_id: the id generated by training service
+            event: the job's state
+            hyper_params: the hyperparameters (a string) generated and returned by tuner
+        """
+        logger.debug('Tuner handle trial end, result is %s', data)
+
+        hyper_params = json_tricks.loads(data['hyper_params'])
+        s, i, _ = hyper_params['parameter_id'].split('_')
+        hyper_configs = self.brackets[int(s)].inform_trial_end(int(i))
+
+        if hyper_configs is not None:
+            logger.debug(
+                'bracket %s next round %s, hyper_configs: %s', s, i, hyper_configs)
+            self.generated_hyper_configs = self.generated_hyper_configs + hyper_configs
+            for _ in range(self.credit):
+                self._request_one_trial_job()
+
+        # Finish this bracket and generate a new bracket
+        if self.brackets[int(s)].no_more_trial:
+            self.curr_s -= 1
+            self.generate_new_bracket()
+        return True
+
+    def handle_report_metric_data(self, data):
+        """reveice the metric data and update Bayesian optimization with final result
+
+        Parameters
+        ----------
+        data:
+            it is an object which has keys 'parameter_id', 'value', 'trial_job_id', 'type', 'sequence'.
+
+        Raises
+        ------
+        ValueError
+            Data type not supported
+        """
+        logger.debug('handle report metric data = %s', data)
+
+        assert 'value' in data
+        value = extract_scalar_reward(data['value'])
+        if self.optimize_mode is OptimizeMode.Maximize:
+            reward = -value
+        else:
+            reward = value
+        assert 'parameter_id' in data
+        s, i, _ = data['parameter_id'].split('_')
+
+        logger.debug('bracket id = %s, metrics value = %s, type = %s', s, value, data['type'])
+        s = int(s)
+
+        assert 'type' in data
+        if data['type'] == 'FINAL':
+            # and PERIODICAL metric are independent, thus, not comparable.
+            assert 'sequence' in data
+            self.brackets[s].set_config_perf(
+                int(i), data['parameter_id'], data['sequence'], value)
+            self.completed_hyper_configs.append(data)
+       
+            _parameters = self.parameters[data['parameter_id']]
+            _parameters.pop(_KEY)
+            # update BO with loss, max_s budget, hyperparameters
+            self.cg.new_result(loss=reward, budget=data['sequence'], parameters=_parameters, update_model=True)
+        elif data['type'] == 'PERIODICAL':
+            self.brackets[s].set_config_perf(
+                int(i), data['parameter_id'], data['sequence'], value)
+        else:
+            raise ValueError(
+                'Data type not supported: {}'.format(data['type']))
+
+        return True
+
+    def handle_add_customized_trial(self, data):
```

Comment:
What is the function of handle_add_customized_trial ?

Reply:
This function is inherited from "MsgDispatcherBase".